keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,657 @@
+strategies to select a flavor.
+
+#### Sorting Workloads that are heads of ClusterQueues
+
+Sorting uses the following criteria:
+
+1. Flavors that don't borrow first.
+2. [New criteria] Highest priority first.
+3. Older creation timestamp first.
+
+Note that these criteria might put Workloads that require preemption ahead,
+because preemption doesn't require borrowing more resources. This is desired,
+because preemption to recover quota or admit high priority Workloads takes
+preference over borrowing.
+
+#### Admission
+
+When iterating over workloads to be admitted, in the order given by the previous
+section, we disallow borrowing in the cohort in the current cycle after
+evaluating a Workload that doesn't require borrowing. This is the same behavior
+that we have today, but note that this criteria includes Workloads that need
+preemption, because there is no preemption with borrowing quota.
+
+This guarantees that, in future cycles, we can admit Workloads that were not
+heads in their ClusterQueues in this cycle, but could fit without borrowing in
+the next cycle, before lending quota to other ClusterQueues.
+
+In the past, we only disallowed borrowing in the cohort if we were able to
+admit the Workload, because we only kept track of flavor assignments of type 0.
+This caused ClusterQueues in the cohort to continue
+borrowing quota, even if there were pending Workloads that would fit under the
+min Quota for their ClusterQueues.
+
+It is actually possible to limit borrowing only for the involved flavors,
+instead of restricting borrowing for all the flavors in the cohort. But we will
+leave this as a future possible optimization to improve throughput.
+
+#### Preemption
+
+For each Workload that got flavor assignments of type 1 or 2, we might need to
+preempt some admitted Workloads.
+
+The algorithm goes like follows:
+
+0. Check that `ClusterQueue.spec.waitBeforePreemptionSeconds` is satisfied.
+   The time is based on the `lastTransitionTime` of the condition `Admitted`
+   (with `status=False`). If it doesn't exist, we use the current time.
+   If the period is not satisfied, we abort preemption.
+   
+   If this is a ClusterQueue with BestEffortFIFO, the Workload is put in the
+   inadmissible staging area with a deadline equal to the remaining time to
+   fullfil `waitBeforePreemptionSeconds`.
+
+   If this is a ClusterQueue with StrictFIFO, the Workload will be requeued
+   and retried immediately in the next admission cycle.
+
+1. Check whether preemption could help.
+
+   Preemption is limited to Workloads with a priority lower than the incoming
+   Workload. To avoid unnecessary preemption calculations, we can keep a
+   priority queue with the priorities of active Workloads in the ClusterQueue
+   and the cohort. If the lowest priority in these sets is higher than or equal
+   to the priority of the incoming Workload, we can skip the preemption
+   algorithm.
```

Comment:
Can a lower priority workload preempt a higher priority one started on borrowed quota? if so, then this optimization will overlook this case, no?

Reply:
Yes, that is true. And I realized that I'm contradicting myself here, with respect to the section "Why no control to opt-out a ClusterQueue from preemption"

I think it makes sense to preempt a higher priority workload in this scenario. If you agree, I can change the details.

Reply:
as discussed offline, lets add an option to the CQ to control this behavior

Reply:
Done